### ISSUE TITLE
Rule extension proposal: id-blacklist accepts regular expressions

### ISSUE THREAD

**Issue by lemonmade** [eslint/eslint#6212]
**What version of ESLint are you using?**

2.10.2

**What problem do you want to solve?**

We are in the process of converting a large CoffeeScript codebase to ES2015+. That conversion generates a lot of temporary variables that we would like developers to provide more meaningful names for before committing. These temporary variables all have the format `ref\d*` (`ref`, `ref1`, `ref2`, etc). I would like to be able to provide a regular expression to the `id-blacklist` rule in order to avoid having to provide a list of all numbered reference names.

**My take on the potential solution**

Allow regular expressions to be passed to the `id-blacklist` rule by surrounding the regular expression in `/` and `/`. This would signal that it should be treated as a regex. We'd then pre-compile those into regular expressions, and would augment the function that checks whether a given identifier name is blacklisted to understand both string and regular expression matches.

I'm happy to take a crack at this if you are interested!

**Comment by ilyavolodin** [comment]
Hm. I thought we already had support for RegExp in `id-blacklist`. I think it's a reasonable enhancement. 👍

**Comment by platinumazure** [comment]
@ilyavolodin Does `id-match` subsume this functionality? (I've been wondering for some time now if we should condense all of those rules into one. I'm not necessarily advocating that we require users to use a massive regular expression to cover all their needs, but rather that we could have rule options in one rule to support lengths, blacklists/whitelists, and regexes.)

**Comment by ilyavolodin** [comment]
If we wouldn't implement the policy of no rule removal, I would say those two should be combined into a single rule `id-name` with option for whitelist and blacklist. But since we are not removing any of the rules, adding RegExp support seems like a reasonable idea. @eslint/eslint-team thoughts?

**Comment by BYK** [comment]
@ilyavolodin that would also imply merging `id-length` into this which I'd support.

**Comment by platinumazure** [comment]
:+1: to merging all the `id-*` rules as a test case for the deprecation plan.

**Comment by mikesherov** [comment]
how would merging id-length work? Would you have to supply a regex like `/.{3,}/`? Or are you just saying merge all the id-* rules into one, and have `length` as an option?

I don't see a clear advantage to merging the rules. 2 rules with no options is as easy to use as 1 rule with 2 options, IMO

**Comment by ilyavolodin** [comment]
@mikesherov I think in general we try to keep rules that are doing the same thing grouped. For example, we used to have `no-semi` and `semi` rules, which we combined into a single rule. A few other examples as well. I'm not 100% sure about combining `id-match` and `id-blacklist` with `id-length`, but the first two are doing more or less the same thing.
If we wanted to combine all three configuration might look something like this:
```js
"id-name": ["error", { "allow": { "pattern": "^[a-z]+([A-Z][a-z]+)*$", "min-length": 3, "max-length": 20 }, "deny": { "blacklist": ["error", "data", "err", "e", "cb", "callback"]}]
```

**Comment by BYK** [comment]
@mikesherov I was suggesting what @ilyavolodin wrote 😀

**Comment by mikesherov** [comment]
Ok, I guess the question I have then is: What's the behavior when someone provides a conflicting allow regex and deny blacklist? Or an allow regex that conflicts with a max or min length?

**Comment by BYK** [comment]
@mikesherov - we still have that conflict, unresolved if you think about it :)

**Comment by ilyavolodin** [comment]
@mikesherov Good question. I would guess we would still report an error. We can't evaluate regex to make sure two of them don't collide, that would be too much trouble. We have to assume that the user knows what they are doing. My only concern about merging those rules together is that we'll have 3 rules hanging around codebase for a year, before they get removed. But that's what we agreed on.
@nzakas What's your feeling about merging all of those rules together?

P.S. Keep in mind, configuration that I provided isn't final by any means. For example, length might need to be moved out into it's own top level property. There's also a switch for `id-match` to check object properties, that's missing from my sample config.

**Comment by nzakas** [comment]
I think we should pull this conversation back to the original request, which is if it makes sense to allow regex patterns in `id-blacklist`. I think it does. 👍 

I don't see a ton of value in merging rules. The rules we have inhabit relatively isolated corners of the problem such that they are fairly easy to use together. I don't see merging rules as improving that situation in any way, so I'm 👎 amd would like us to refocus on the original request here.

**Comment by mikesherov** [comment]
I'm in the exact same camp as @nzakas here. :+1: to proposal. :-1: on merge.

**Comment by kaicataldo** [comment]
Agreed with @nzakas and @mikesherov. I can see the case for merging `id-match` and `id-blacklist`, but I don't see the benefit in merging id-length in as well. Either way, I think we should stick to the proposal at hand, and if people feel really strongly about wanting to merge `id-match` and `id-blacklist` we can discuss that in a new issue.

**Comment by kaicataldo** [comment]
So, to be clear, :+1: for the enhancement

**Comment by ilyavolodin** [comment]
Marking as accepted since we have more then enough votes.

**Comment by kaicataldo** [comment]
Closing this as there hasn't been enough interest to drive this to completion. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.
